src/shared/tpm2-util.c

```diff
@@ -1709,87 +1762,36 @@ static int find_signature(
 #endif
 }
 
-static int tpm2_make_policy_session(
+static int tpm2_build_sealing_policy(
                 Tpm2Context *c,
-                const Tpm2Handle *primary,
-                const Tpm2Handle *parent_session,
-                TPM2_SE session_type,
+                const Tpm2Handle *session,
                 uint32_t hash_pcr_mask,
-                uint16_t pcr_bank, /* If UINT16_MAX, pick best bank automatically, otherwise specify bank explicitly. */
+                uint16_t pcr_bank,
                 const void *pubkey,
                 size_t pubkey_size,
                 uint32_t pubkey_pcr_mask,
                 JsonVariant *signature_json,
                 bool use_pin,
-                Tpm2Handle **ret_session,
-                TPM2B_DIGEST **ret_policy_digest,
-                TPMI_ALG_HASH *ret_pcr_bank) {
+                TPM2B_DIGEST **ret_policy_digest) {
 
-        static const TPMT_SYM_DEF symmetric = {
-                .algorithm = TPM2_ALG_AES,
-                .keyBits.aes = 128,
-                .mode.aes = TPM2_ALG_CFB,
-        };
         TSS2_RC rc;
         int r;
 
         assert(c);
+        assert(session);
         assert(pubkey || pubkey_size == 0);
         assert(pubkey_pcr_mask == 0 || pubkey_size > 0);
 
-        log_debug("Starting authentication session.");
-
-        /* So apparently some TPM implementations don't implement trial mode correctly. To avoid issues let's
```

Comment:
here it says that trial mode is flaky, but now it's always used, is that going to cause issues?

Reply:
well, i actually suspect the problem wasn't the tpm, because our code was wrong; we were attempting to perform PolicyPCR but we weren't reading the PCR values first, and providing those to the PolicyPCR call (as the expected values). The spec says that when this is done *in trial mode*, the tpm implementation *may* choose to read the current PCR values and use those, but it doesn't have to; it's acceptable for the tpm to return error instead. My assumption is that's what some tpms did, which was (incorrectly) interpreted as not implementing trial mode correctly.

> If this command is used for a trial policySession, policySession→policyDigest will be updated using the
> values from the command rather than the values from a digest of the TPM PCR. If the caller does not
> provide PCR settings (pcrDigest has a length of zero), the TPM may (and it is preferred to) use the
> current TPM PCR settings (digestTPM) in the calculation for the new policyDigest. The TPM may return
> an error if the caller does not provide a PCR digest for a trial policy session but this is not the preferred
> behavior.
> 

Of course, I don't know the specific cause of the statement, I'm just guessing; but in any case, the next PR in this larger series adds policy calculation so we can avoid trial sessions entirely. I'll add back in the comment and check, and then remove it later when we move over to using calculated values instead of a trial session.

Reply:
specifically #27517 replaces the trial policy with calculations for the policy hash.

Reply:
Can you add this to the commit message? It's useful information to record

Reply:
yep, updated the commit msg and pushed